src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs

```diff
@@ -45,60 +57,224 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
+                // NoPia shortcut either didn't apply or failed, go through general process 
+                // of matching candidates.
 
-            // Based on assembly identity, for each assembly, 
-            // bind its references against the other assemblies we have.
-            for (int i = 0; i < totalAssemblies; i++)
+                ReuseAssemblySymbols(boundInputs, candidateInputAssemblySymbols, assemblies, corLibraryIndex);
+
+                return boundInputs;
+            }
+            finally
             {
-                boundInputs[i].ReferenceBinding = assemblies[i].BindAssemblyReferences(assemblies, this.IdentityComparer);
+                referenceBindings.Free();
             }
+        }
 
-            // All assembly symbols should be uninitialized at this point:
-            Debug.Assert(Array.TrueForAll(boundInputs, bi => bi.AssemblySymbol == null));
+        private void ResolveAndBindMissingAssemblies(
+            ImmutableArray<AssemblyData> explicitAssemblies,
+            ArrayBuilder<AssemblyReferenceBinding[]> referenceBindings,
+            ArrayBuilder<MetadataReference> resolvedReferences,
+            MetadataReferenceResolver resolver,
+            MetadataImportOptions importOptions,
+            DiagnosticBag resolutionDiagnostics,
+            out ImmutableArray<AssemblyData> allAssemblies)
+        {
+            var implicitAssemblies = ArrayBuilder<AssemblyData>.GetInstance();
 
-            hasCircularReference = CheckCircularReference(boundInputs);
+            // tracks identities we already asked the resolver to resolve:
+            var resolvedMissingIdentities = PooledHashSet<AssemblyIdentity>.GetInstance();
 
-            corLibraryIndex = IndexOfCorLibrary(assemblies);
+            // reference bindings of implicit assemblies, use to calculate a fixed point:
+            var referenceBindingsToProcess = ArrayBuilder<AssemblyReferenceBinding[]>.GetInstance();
 
-            // For each assembly, locate AssemblySymbol with similar reference resolution
-            // What does similar mean?
-            // Similar means: 
-            // 1) The same references are resolved against the assemblies that we are given.
-            // 2) The same assembly is used as the COR library.
+            try
+            {
+                // collect all missing identities, resolve the assemblies and bind their references against explicit definitions:
+                referenceBindingsToProcess.AddRange(referenceBindings);
 
-            TAssemblySymbol[] candidateInputAssemblySymbols = new TAssemblySymbol[totalAssemblies];
+                while (referenceBindingsToProcess.Count > 0)
+                {
+                    foreach (var binding in referenceBindingsToProcess.Pop())
+                    {
+                        // try resolve a reference if eitehr the binding failed or didn't find an exact match:
+                        if (binding.IsBound && binding.VersionDifference == 0 || 
```

Comment:
Should binding to an explicitly provided assembly take priority, even when match is inexact? After all this is what the user explicitly chose to use.